freedomofpress/securedrop-client
# app: add pending reply status, persist replies in the database

# Description

Fixes #350. 
Fixes #294. 

What it looks like now when you send replies (the failed replies will persist between application restarts and clicking between sources):

![pending_status](https://user-images.githubusercontent.com/7832803/67238046-76256500-f41a-11e9-956d-2725b180390f.gif)

For a followup:
* Add spinner: https://app.zeplin.io/project/5c807ea562f734bd2756b243/screen/5d5b183a152c905223e5fab0 - will file in a followup issue since we don't actually have a nice place in the conversation item widget to put this yet and I don't want this diff to get any larger (actually there already is an issue - #359)

# Test Plan

To generate lots of failed replies and introduce delay I'm testing using the following server diff:

```
diff --git a/securedrop/journalist_app/api.py b/securedrop/journalist_app/api.py
index aa776d520..8eb8f26c8 100644
--- a/securedrop/journalist_app/api.py
+++ b/securedrop/journalist_app/api.py
@@ -230,6 +230,13 @@ def make_blueprint(config):
                 {'replies': [reply.to_json() for
                              reply in source.replies]}), 200
         elif request.method == 'POST':
+            import time
+            time.sleep(2)
+
+            import random
+            fail_early = random.choice([True, False])
+            if fail_early:
+                abort(409, 'That UUID is already in use.')
             source = get_or_404(Source, source_uuid,
                                 column=Source.uuid)
             if request.json is None:
```

You can use that or you can just use staging and rely on regular ol tor network issues to:

1. Send a reply (or a few) to a source (source A) until one fails.
2. Click to another source (source B).
3. Click back to the source A. Confirm that the failed reply is still there.
4. Close the application.
5. Restart the application with the same sdc homedir. Confirm the failed reply is still there. 

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:
